File: smash-ranks-android/app/src/main/java/com/garpr/android/views/MatchItemView.java
package com.garpr.android.views;

import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.os.Build;
import android.support.annotation.AttrRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StyleRes;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.view.View;
import android.widget.TextView;

import com.garpr.android.App;
import com.garpr.android.R;
import com.garpr.android.activities.PlayerActivity;
import com.garpr.android.adapters.BaseAdapterView;
import com.garpr.android.misc.FavoritePlayersManager;
import com.garpr.android.misc.MiscUtils;
import com.garpr.android.models.Match;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MatchItemView extends IdentityFrameLayout implements BaseAdapterView<Match>,
        View.OnClickListener, View.OnLongClickListener {

    private Match mContent;

    @Inject
    FavoritePlayersManager mFavoritePlayersManager;

    @BindView(R.id.tvName)
    TextView mName;


    public MatchItemView(@NonNull final Context context, @Nullable final AttributeSet attrs) {
        super(context, attrs);
    }

    public MatchItemView(@NonNull final Context context, @Nullable final AttributeSet attrs,
            @AttrRes final int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public MatchItemView(@NonNull final Context context, @Nullable final AttributeSet attrs,
            @AttrRes final int defStyleAttr, @StyleRes final int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    public Match getContent() {
        return mContent;
    }

    @Override
    protected String getIdentityId() {
        return mContent.getOpponent().getId();
    }

    @Override
    protected void identityIsSomeoneElse() {
        super.identityIsSomeoneElse();
        styleTextViewForSomeoneElse(mName);
    }

    @Override
    protected void identityIsUser() {
        super.identityIsUser();
        styleTextViewForUser(mName);
    }

    @Override
    public void onClick(final View v) {
        final Context context = getContext();
        final Activity activity = MiscUtils.optActivity(context);

        if (activity instanceof OnClickListener) {
            ((OnClickListener) activity).onClick(this);
        } else {
            context.startActivity(PlayerActivity.getLaunchIntent(context,
                    mContent.getOpponent().getId(), mContent.getOpponent().getName()));
        }
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        ButterKnife.bind(this);
        setOnClickListener(this);
        setOnLongClickListener(this);

        if (isInEditMode()) {
            return;
        }

        App.get().getAppComponent().inject(this);
    }

    @Override
    public boolean onLongClick(final View v) {
        return mFavoritePlayersManager.showAddOrRemovePlayerDialog(getContext(),
                mContent.getOpponent());
    }

    @Override
    public void setContent(final Match content) {
        mContent = content;

        mName.setText(mContent.getOpponent().getName());

        switch (mContent.getResult()) {
            case EXCLUDED:
                mName.setTextColor(MiscUtils.getAttrColor(getContext(),
                        android.R.attr.textColorSecondary));
                break;

            case LOSE:
                mName.setTextColor(ContextCompat.getColor(getContext(), R.color.lose));
                break;

            case WIN:
                mName.setTextColor(ContextCompat.getColor(getContext(), R.color.win));
                break;

            default:
                throw new RuntimeException("unknown result: " + mContent.getResult());
        }

        refreshIdentity();
    }


    public interface OnClickListener {
        void onClick(final MatchItemView v);
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/views/MatchItemView.kt
package com.garpr.android.views

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.garpr.android.R
import com.garpr.android.activities.PlayerActivity
import com.garpr.android.adapters.BaseAdapterView
import com.garpr.android.extensions.activity
import com.garpr.android.extensions.appComponent
import com.garpr.android.extensions.clear
import com.garpr.android.extensions.getAttrColor
import com.garpr.android.managers.FavoritePlayersManager
import com.garpr.android.managers.RegionManager
import com.garpr.android.models.Match
import com.garpr.android.models.MatchResult
import kotterknife.bindView
import javax.inject.Inject

class MatchItemView @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null
) : IdentityFrameLayout(context, attrs), BaseAdapterView<Match>, View.OnClickListener,
        View.OnLongClickListener {

    @Inject
    protected lateinit var favoritePlayersManager: FavoritePlayersManager

    @Inject
    protected lateinit var regionManager: RegionManager

    private val name: TextView by bindView(R.id.tvName)


    interface OnClickListener {
        fun onClick(v: MatchItemView)
    }

    override fun clear() {
        super.clear()

        name.clear()
        refresh()
    }

    override fun identityIsSomeoneElse() {
        super.identityIsSomeoneElse()
        styleTextViewForSomeoneElse(name)
    }

    override fun identityIsUser() {
        super.identityIsUser()
        styleTextViewForUser(name)
    }

    var match: Match? = null
        private set(value) {
            field = value

            if (value == null) {
                clear()
                return
            }

            identity = value.opponent
            name.text = value.opponent.name

            when (value.result) {
                MatchResult.EXCLUDED -> name.setTextColor(context.getAttrColor(android.R.attr.textColorSecondary))
                MatchResult.LOSE -> name.setTextColor(ContextCompat.getColor(context, R.color.lose))
                MatchResult.WIN -> name.setTextColor(ContextCompat.getColor(context, R.color.win))
            }

            refresh()
        }

    override fun onClick(v: View) {
        val match = this.match ?: return
        val activity = this.activity

        if (activity is OnClickListener) {
            activity.onClick(this)
        } else {
            val opponent = match.opponent
            context.startActivity(PlayerActivity.getLaunchIntent(context, opponent.id,
                    regionManager.getRegion(context)))
        }
    }

    override fun onFinishInflate() {
        super.onFinishInflate()

        if (!isInEditMode) {
            appComponent.inject(this)
        }

        setOnClickListener(this)
        setOnLongClickListener(this)
    }

    override fun onLongClick(v: View): Boolean {
        return favoritePlayersManager.showAddOrRemovePlayerDialog(context, match?.opponent,
                regionManager.getRegion(context))
    }

    override fun setContent(content: Match) {
        match = content
    }

}
